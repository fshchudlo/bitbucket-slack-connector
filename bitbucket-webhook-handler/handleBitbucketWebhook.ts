import * as useCases from "./use-cases";
import { SlackAPIAdapter } from "./ports/SlackAPIAdapter";
import { BitbucketAPIAdapter } from "./ports/BitbucketAPIAdapter";
import { BitbucketNotification } from "../bitbucket-payload-types";
import { WebhookHandlerConfig } from "./webhookHandlerConfig";
import { provisionPullRequestChannel } from "./provisionPullRequestChannel";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketAPI: BitbucketAPIAdapter, config: WebhookHandlerConfig) {
    const eventKey = payload.eventKey;
    const channelInfo = await provisionPullRequestChannel(slackAPI, bitbucketAPI, payload, config);

    switch (eventKey) {
        case "pr:opened":
            await useCases.inviteParticipantsAndSetChannelBookmark(payload, slackAPI, config.defaultChannelParticipants, channelInfo.id);
            await useCases.tryBroadcastMessageAboutOpenedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        case "pr:modified":
            await useCases.sendMessageAboutPRModification(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:updated":
            await useCases.updateChannelMembers(payload, slackAPI, channelInfo.id);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await useCases.sendMessageAboutReviewerAction(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:added":
            await useCases.sendMessageAboutAddedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:edited":
            await useCases.sendMessageAboutEditedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:comment:deleted":
            await useCases.sendMessageAboutDeletedComment(payload, slackAPI, channelInfo.id);
            break;
        case "pr:from_ref_updated":
            await useCases.sendMessageAboutNewCommit(payload, slackAPI, bitbucketAPI, channelInfo.id);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await useCases.sendCompletionMessageAndCloseTheChannel(payload, slackAPI, channelInfo.id);
            await useCases.tryBroadcastMessageAboutClosedPR(payload, slackAPI, config.getOpenedPRBroadcastChannelId(payload));
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}

