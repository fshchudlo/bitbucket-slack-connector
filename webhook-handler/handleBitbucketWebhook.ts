import {
    sendCompletionMessageAndCloseTheChannel,
    createChannelAndInviteParticipants,
    sendMessageAboutAddedComment,
    updateChannelMembers,
    sendMessageAboutNewCommit,
    sendMessageAboutReviewerAction,
    sendMessageAboutDeletedComment, sendMessageAboutEditedComment
} from "./use-cases";
import { SlackAPIAdapter } from "./SlackAPIAdapter";
import { BitbucketGateway } from "./BitbucketGateway";
import { sendMessageAboutPRModification } from "./use-cases/sendMessageAboutPRModification";
import { BitbucketNotification } from "../typings";

export default async function handleBitbucketWebhook(payload: BitbucketNotification, slackAPI: SlackAPIAdapter, bitbucketGateway: BitbucketGateway, usePrivateChannels: boolean = true) {
    const eventKey = payload.eventKey;
    switch (eventKey) {
        case "pr:opened":
            await createChannelAndInviteParticipants(payload, slackAPI, usePrivateChannels);
            break;
        case "pr:reviewer:unapproved":
        case "pr:reviewer:needs_work":
        case "pr:reviewer:approved":
            await sendMessageAboutReviewerAction(payload, slackAPI);
            break;
        case "pr:comment:added":
            await sendMessageAboutAddedComment(payload, slackAPI);
            break;
        case "pr:comment:edited":
            await sendMessageAboutEditedComment(payload, slackAPI);
            break;
        case "pr:comment:deleted":
            await sendMessageAboutDeletedComment(payload, slackAPI);
            break;
        case "pr:from_ref_updated":
            await sendMessageAboutNewCommit(payload, slackAPI, bitbucketGateway);
            break;
        case "pr:reviewer:updated":
            await updateChannelMembers(payload, slackAPI);
            break;
        case "pr:modified":
            await sendMessageAboutPRModification(payload, slackAPI);
            break;
        case "pr:merged":
        case "pr:declined":
        case "pr:deleted":
            await sendCompletionMessageAndCloseTheChannel(payload, slackAPI);
            break;
        default:
            throw new Error(`"${eventKey}" event key is unknown.`);
    }
}