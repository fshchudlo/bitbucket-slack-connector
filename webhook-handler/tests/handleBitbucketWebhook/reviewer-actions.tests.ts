import SlackAdapterSnapshottingMock from "../SlackAdapterSnapshottingMock";
import TestPayloadBuilder from "../TestPayloadBuilder";
import handleBitbucketWebhook from "../../handleBitbucketWebhook";
import { TestBitbucketGateway } from "../TestBitbucketGateway";

describe("handleBitbucketWebhook", () => {
    it("Should send message on PR approval/unapproval/needs work", async () => {
        const testSlackGateway = new SlackAdapterSnapshottingMock();

        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestNeedsWork(), testSlackGateway, new TestBitbucketGateway());
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestUnapproved(), testSlackGateway, new TestBitbucketGateway());
        await handleBitbucketWebhook(TestPayloadBuilder.pullRequestApproved(), testSlackGateway, new TestBitbucketGateway());

        expect(testSlackGateway.snapshot).toMatchSnapshot();
    });
});
