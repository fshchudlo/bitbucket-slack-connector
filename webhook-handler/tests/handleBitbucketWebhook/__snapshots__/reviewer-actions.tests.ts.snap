// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handleBitbucketWebhook Should send message on PR approval/unapproval/needs work 1`] = `
{
  "archivedChannels": [],
  "createdChannels": [],
  "invitesToChannels": [],
  "kicksFromChannels": [],
  "lookedUpUsers": [],
  "searchedChannels": [
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
    {
      "channelName": "pr-test-proj-test-repository-1",
      "excludeArchived": true,
    },
  ],
  "searchedCommentSnapshots": [],
  "sentMessages": [
    {
      "blocks": [
        {
          "text": {
            "text": "Test Reviewer 1 requested changes for <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": ":large_yellow_circle:  Unapproved: Test Reviewer 1",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": "Test Reviewer 1 requested changes for <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
    {
      "blocks": [
        {
          "text": {
            "text": "Test Reviewer 1 unapproved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": ":large_yellow_circle:  Unapproved: Test Reviewer 1",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": "Test Reviewer 1 unapproved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
    {
      "blocks": [
        {
          "text": {
            "text": "Test Reviewer 1 approved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
            "type": "mrkdwn",
          },
          "type": "section",
        },
        {
          "text": {
            "text": ":large_yellow_circle:  Unapproved: Test Reviewer 1",
            "type": "mrkdwn",
          },
          "type": "section",
        },
      ],
      "channelId": "12345",
      "iconEmoji": ":bitbucket:",
      "text": "Test Reviewer 1 approved <https://git.test.com/projects/TEST-PROJ/repos/test-repository/pull-requests/1/overview|pull request>.",
    },
  ],
  "setChannelTopics": [],
}
`;
